File: Practice/Dynamic Programming/LetsParty.js
function LetsParty(A) {
  let arr = [];

  for (let i = 0; i <= A; i++) {
    arr.push(-1);
  }

  let ans = calculate(A);
  let mod = 10003;

  return ans % mod;
  function calculate(A) {
    if (A == 1) return 1;
    if (A == 2) return 2;

    if (arr[A] != -1) return arr[A];

    arr[A] = calculate(A - 1) + calculate(A - 2);

    return arr[A];
  }
}
A = 5;
console.log(LetsParty(A));

File: Revision/Dynamic Programming/LetsParty.java
import java.util.Arrays;

public class LetsParty {
    public static void main(String[] args) {
        int A = 4;
        int mod = 10003;
        int[] dp = new int[A + 1];
        Arrays.fill(dp, -1);

        LetsParty letsParty = new LetsParty();
        int ans = letsParty.noOfWays(A, dp, mod);

        System.out.println(ans);

    }

    public int noOfWays(int A, int[] dp, int mod) {
        if (A < 3)
            return A;

        if (dp[A] != -1)
            return dp[A];
        dp[A] = (noOfWays(A - 1, dp, mod) + (noOfWays(A - 2, dp, mod) * (A - 1)) % mod) % mod;
        return dp[A];
    }
}
